### ISSUE TITLE
storpool-block charm requires hard-coded cinder hostnames as charm config

### ISSUE THREAD

**Issue by ryan-beisner** [storpool/charm-storpool-block#4]
The storpool-block charm requires hard-coded cinder hostnames as charm config.  This is problematic to any sort of automated CI.  The hostnames are not predictable and are not known until mid-deployment.

```
cs:~openstack-dev-s/xenial/storpool-block

The storpool_conf option needs the following:

      storpool_conf: |
        SP_CLUSTER_NAME=Canonical-PoC-VIRT1
        SP_CLUSTER_ID=abd.b
        SP_IFACE=eno2
        SP_API_HTTP_HOST=192.168.36.201
        SP_API_HTTP_PORT=81
        SP_AUTH_TOKEN=3050623014048672255
        SP_EXPECTED_NODES=1
        SP_NODE_NON_VOTING=1
        SP_USE_CGROUPS=1
        [hostname_of_cinder1]    <============= hardcoded hostname here
        SP_OURID=11
        [hostname_of_cinder2]   <============= hardcoded hostname here
        SP_OURID=12
        [hostname_of_cinder3]    <============= hardcoded hostname here
        SP_OURID=13
```

**Comment by ppentchev** [comment]
Hi Ryan,

TL;DR: we should figure out some way to get the CI to use the configuration generated by the storpool-charms tool or, in a couple of days, the library that the tool is turning into as we speak.

Well, the truth is that this is the way the StorPool services (including the StorPool client) work: they expect to find an /etc/storpool.conf file that contains some configuration settings that are the same for all the nodes in the cluster and others that are specific to this particular node.  Each node needs a numeric identifier between 1 and 64 that is unique for the cluster, and the services running on the node need to know which network interfaces they should communicate with the StorPool cluster on.  Generally, the expectation is that the StorPool configuration will be crafted carefully for the whole cluster to ensure maximum reliability and performance; there are actually many more options that the charm glosses over and leaves at their default values.

Honestly, I think that it would be quite hard to get the charm to configure this automatically; I have thought about this more than a little, and while the network interface may be picked up by passing a network space as a parameter, figuring out a unique numeric ID would be a problem if not all units are up yet or if some of them are currently misbehaving in any way.  There is the obvious "derive it from the unit number" idea, but I don't really see an easy way that would work given that the unit numbers increase monotonically to, well, quite large values, and wrapping them using e.g. a modulo operation might cause Bad Things(tm) to happen if, say, one unit is stable and another one is constantly being removed and readded.

Last year I actually came up with a solution to this problem, at least for a setup where the OpenStack charms are already installed: run something on a server that has the right to issue "juju" commands (e.g. `juju --format=json status` and `juju ssh 23 systemctl --no-pager list-unit-files`), gather some information about the currently deployed charms, figure out which nodes run Nova, which ones run Cinder, etc, and generate a configuration that encompasses all that.  This is the storpool-charms tool: https://github.com/storpool/storpool-charms - its `generate-config` and `generate-charm-config` commands should work as a starting point.

For the CI case, deploying the storpool-block charm with partial configuration (the PPA, a couple of version settings, etc), then waiting for the OpenStack charms to be deployed and the nodes to come up, and then running `storpool-charms generate-config` should provide some indication as to what to use as the "storpool_conf" setting.  It is not completely usable: some of the general settings would be wrong for the CI cluster, such as the StorPool cluster name or the expected number of server nodes (at least 3 in a normal cluster, but only 1 (the StorPool appliance) for the CI setup), IP addresses, etc.  But the node-specific sections should be usable.

So right now you could try to do something with the storpool-charms output and some creative use of awk, sed, or something else that filters out anything before the first section header.  I am currently working on extending the storpool-charms tool, turning it into a library of modules, so that it may be used more easily and provide just the information as, say, a Python dictionary, instead of a full configuration file.

Let me know your thoughts on this, and we can hopefully find a way forward.

Best regards,
Peter

**Comment by ppenchev-storpool** [comment]
Hi Ryan,

So it was a bit of a larger change than I initially set out to do, but it's ready now - the `storpool-charms.py` command-line tool has been turned into the `storpool.charms.manage` library (with a `spcharms_manage` command-line entry point), converted to Python 3 only, had unit tests added, had type annotations added, and had some minor buglets fixed.

My suggestion is to have, as part of the CI deployment, the following sequence of events:
- install most of the Juju charms, including the `cinder-storpool` and `storpool-block` ones, but without configuration for `storpool-block`
- download and install the `storpool_charms_manage` module from PyPI (or, if you'd like, I could package it up and put it in our PPA or in another, public PPA of ours)
- use the `get_storpool_config_data()` function from the `storpool.charms.manage.juju` module

Something like the following should work:

    from storpool.charms.manage import config as sconfig
    from storpool.charms.manage import juju as sjuju

    cfg = sconfig.Config(space='storpool', repo_auth='username:password')
    status = sjuju.get_status(cfg)

    spcfg = sjuju.get_storpool_config_data(cfg, status)
    for hostname, data in spcfg.items():
        print('Host {name}: StorPool ID {oid}'
              .format(name=hostname, oid=data['SP_OURID']))

The `get_storpool_config_data()` function returns a dictionary with hostname/config pairs; the config is a name/string-value dictionary that currently sets the `SP_OURID` and `SP_IFACE` variables for StorPool to use.  There is also the `get_storpool_config()` function in the same namespace which may demonstrate how to turn the dictionary into the per-machine sections of the `storpool.conf` file, but, as noted in my previous reply, it is not directly usable for a CI since it adds some sample cluster configuration before that.  This, combined with the rest of the charm configuration that we have already discussed with Ashley, should help get even a more generic CI system running.

BTW it seems that I sent my first reply to this issue from my personal GitHub account; sorry about that :)

Best regards,
Peter
